docs/development/test-guidelines.adoc

```diff
@@ -0,0 +1,209 @@
+Nevertheless we will present mechanisms to handle this as well.
+
+[[lib-dot-go-source]]
+[source,go]
+.$GOPATH/src/foo/bar/lib.go
+----
+package bar // <1>
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Say returns the input string as is
+func Say(message string) string { // <2>
+	return fmt.Sprintf("%s", message)
+}
+
+// Shout returns the input string in upper case
+// letters with three exlamation marks.
+func Shout(message string) string { // <3>
+	return fmt.Sprintf("%s!!!", toUpper(message))
+}
+
+// toUpper returns the input string in upper case
+func toUpper(message string) string { // <4>
+	return strings.ToUpper(message)
+}
+----
+<1> The package name is `bar`. 
+<2> An exported function called `Say` that is unknown packages that import `bar`.
+<3> Another exported function called `Shout` that is unknown packages that import `bar`.
+<4> A function called `toUpper` that is unexported and unknown packages that import `bar`.
+
+=== Blackbox vs. Whitebox tests
+
+In a blackbox test, you test a package as if you **only have access to the
+exported symbols**. In our <<lib-dot-go-source>> example above we have two exported
+symbols: `Say` and `Shout`. The `toUpper` function is an unexported symbol due
+to its lowercase name. 
+
+A whitebox test has access to all symbols, and not only to the expored ones.
+
+In principle, all symbols in a file can be *seen* by all files in the same package.
+As a consequence, to have a blackbox test, we must write it in a different package
+than `bar`. But in order for `go test` to pick up a test we need to place it in the
+same directory as the <<lib-dot-go-source>> file.
+
+In Go your test filenames must end with `_test.go`. In order to distinguish black-
+and whitebox test files, a recommendation is to use the suffixes `_blackbox_test.go`
+and `_whitebox_test.go`.
+
+[[lib-blackbox-test-dot-go-source]]
+[source,go]
+.$GOPATH/src/foo/bar/lib_blackbox_test.go
+----
+package bar_test // <1>
+
+import (
+	. "foo/bar" // <2>
+	"testing"
+)
+
+func TestSay(t *testing.T) { // <3>
+	t.Parallel() // <4>
+	if Say("heLLo") != "heLLo" { 
+		t.Errorf("Wrong say result")
+	}
+}
+
+func TestShout(t *testing.T) { // <5>
+	t.Parallel() // <6>
+	if Shout("heLLo") != "HELLO!!!" {
+		t.Errorf("Wrong shout result")
+	}
+}
+----
+<1> Notice that Go **allows** the package name `bar_test` to exist in the `bar` folder.
+<2> One of the very few circumstances in which a dot-import is okay in Go. Here it simply
+    allows for easier writing.
+<3> Testing the exported `Say` function.
+<4> Marking the test for the `Say` function test as: "executable in parallel".
+<5> Testing the exported `Shout` function.
+<6> Marking the test for the `Shout` function test as: "executable in parallel".
+
+A few things to notice in <<lib-blackbox-test-dot-go-source>> is the package name: `bar_test`.
+This is an exception to the Go compiler because normally only package names with the same name are
```

Comment:
I did not know that. To be honest: special casing this in the compiler to allow for cross-package tests seems like the wrong fix to me. But if this is the way it's meant to be, let's not try to swim against the stream.